"make bench" fails because metering_benchmark has no compiler specified

When I run:

```
./build make bench
```

I get the following error:

```
     Running target/release/deps/metering_benchmark-771e924d5f51f188
Benchmarking Meter/No Metering: Warming up for 3.0000 sthread 'main' panicked at 'compiler not specified, activate a compiler via features', lib/middleware-common/benches/metering_benchmark.rs:168:5
```

Looking at the code, looks like none of the features were enabled:

```rust
#[cfg(not(any(feature = "llvm", feature = "clif", feature = "singlepass")))]
fn get_compiler(_limit: u64, metering: bool) -> impl Compiler {
    panic!("compiler not specified, activate a compiler via features");
    use wasmer_clif_backend::CraneliftCompiler;
    CraneliftCompiler::new()
}
```

If I try to enable the "backend-llvm" feature explicitly:

```
./build cargo bench --all --features backend-llvm
```

the benchmark fails in the exact same way.

I also tried to enable "llvm" feature:

```
./build cargo bench --all --features llvm
```

but it also fails:

```
error: Package `wasmer v0.5.7 (/wasmer)` does not have these features: `llvm`
```

Is the code broken or did I miss some magic to get `make bench` running?

@MarkMcCaskey Thanks for the clarification!

I don't see these new build targets in `master`, though:

```
$ make bench-clif
make: *** No rule to make target 'bench-clif'.  Stop.
$ make bench-llvm
make: *** No rule to make target 'bench-llvm'.  Stop.
$ make bench-singlepass
make: *** No rule to make target 'bench-singlepass'.  Stop.
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -115,6 +115,8 @@ jobs:
           name: Check
           command: |
             make check
+            make compile-bench-singlepass
+            # TODO: add compile-bench-llvm and compile-bench-clif when they work
       - run:
           name: Release
           command: make release-fast

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -73,9 +73,23 @@ trace = ["wasmer-runtime-core/trace"]
 extra-debug = ["wasmer-clif-backend/debug", "wasmer-runtime-core/debug"]
 # This feature will allow cargo test to run much faster
 fast-tests = []
-backend-cranelift = ["wasmer-runtime-core/backend-cranelift", "wasmer-runtime/cranelift"]
-backend-llvm = ["wasmer-llvm-backend", "wasmer-runtime-core/backend-llvm", "wasmer-runtime/llvm"]
-backend-singlepass = ["wasmer-singlepass-backend", "wasmer-runtime-core/backend-singlepass", "wasmer-runtime/singlepass"]
+backend-cranelift = [
+    "wasmer-runtime-core/backend-cranelift",
+    "wasmer-runtime/cranelift",
+    "wasmer-middleware-common/clif"
+]
+backend-llvm = [
+    "wasmer-llvm-backend",
+    "wasmer-runtime-core/backend-llvm",
+    "wasmer-runtime/llvm",
+    "wasmer-middleware-common/llvm"
+]
+backend-singlepass = [
+    "wasmer-singlepass-backend",
+    "wasmer-runtime-core/backend-singlepass",
+    "wasmer-runtime/singlepass",
+    "wasmer-middleware-common/singlepass"
+]
 wasi = ["wasmer-wasi"]
 # vfs = ["wasmer-runtime-abi"]
 

--- a/Makefile
+++ b/Makefile
@@ -148,6 +148,14 @@ bench-clif:
 bench-llvm:
 	cargo bench --all --no-default-features --features "backend-llvm"
 
+# compile but don't run the benchmarks
+compile-bench-singlepass:
+	cargo bench --all --no-run --no-default-features --features "backend-singlepass"
+compile-bench-clif:
+	cargo bench --all --no-run --no-default-features --features "backend-clif"
+compile-bench-llvm:
+	cargo bench --all --no-run --no-default-features --features "backend-llvm"
+
 
 # Build utils
 build-install:

--- a/Makefile
+++ b/Makefile
@@ -141,8 +141,12 @@ release-singlepass:
 release-llvm:
 	cargo build --release --features backend-llvm
 
-bench:
-	cargo bench --all
+bench-singlepass:
+	cargo bench --all --no-default-features --features "backend-singlepass"
+bench-clif:
+	cargo bench --all --no-default-features --features "backend-clif"
+bench-llvm:
+	cargo bench --all --no-default-features --features "backend-llvm"
 
 
 # Build utils

--- a/README.md
+++ b/README.md
@@ -220,7 +220,10 @@ Each integration can be tested separately:
 Benchmarks can be run with:
 
 ```sh
-make bench
+make bench-[backend]
+
+# for example
+make bench-singlepass
 ```
 
 ## Roadmap

--- a/lib/middleware-common/benches/metering_benchmark.rs
+++ b/lib/middleware-common/benches/metering_benchmark.rs
@@ -164,15 +164,11 @@ fn get_compiler(limit: u64, metering: bool) -> impl Compiler {
 }
 
 #[cfg(not(any(feature = "llvm", feature = "clif", feature = "singlepass")))]
-fn get_compiler(_limit: u64, metering: bool) -> impl Compiler {
-    panic!("compiler not specified, activate a compiler via features");
-    use wasmer_clif_backend::CraneliftCompiler;
-    CraneliftCompiler::new()
-}
+compile_error!("compiler not specified, activate a compiler via features");
 
 #[cfg(feature = "clif")]
 fn get_compiler(_limit: u64, metering: bool) -> impl Compiler {
-    panic!("cranelift does not implement metering");
+    compile_error!("cranelift does not implement metering");
     use wasmer_clif_backend::CraneliftCompiler;
     CraneliftCompiler::new()
 }

--- a/lib/middleware-common/src/metering.rs
+++ b/lib/middleware-common/src/metering.rs
@@ -162,15 +162,11 @@ mod tests {
     }
 
     #[cfg(not(any(feature = "llvm", feature = "clif", feature = "singlepass")))]
-    fn get_compiler(_limit: u64) -> impl Compiler {
-        panic!("compiler not specified, activate a compiler via features");
-        use wasmer_clif_backend::CraneliftCompiler;
-        CraneliftCompiler::new()
-    }
+    compile_error!("compiler not specified, activate a compiler via features");
 
     #[cfg(feature = "clif")]
     fn get_compiler(_limit: u64) -> impl Compiler {
-        panic!("cranelift does not implement metering");
+        compile_error!("cranelift does not implement metering");
         use wasmer_clif_backend::CraneliftCompiler;
         CraneliftCompiler::new()
     }

--- a/src/bin/wasmer.rs
+++ b/src/bin/wasmer.rs
@@ -41,7 +41,12 @@ mod wasmer_wasi {
         false
     }
 
-    pub fn generate_import_object(_args: Vec<Vec<u8>>, _envs: Vec<Vec<u8>>) -> ImportObject {
+    pub fn generate_import_object(
+        _args: Vec<Vec<u8>>,
+        _envs: Vec<Vec<u8>>,
+        _preopened_files: Vec<String>,
+        _mapped_dirs: Vec<(String, std::path::PathBuf)>,
+    ) -> ImportObject {
         unimplemented!()
     }
 }
@@ -620,11 +625,14 @@ fn execute_wasm(options: &Run) -> Result<(), String> {
                 if let Err(ref err) = result {
                     match err {
                         RuntimeError::Trap { msg } => panic!("wasm trap occured: {}", msg),
+                        #[cfg(feature = "wasi")]
                         RuntimeError::Error { data } => {
                             if let Some(error_code) = data.downcast_ref::<wasmer_wasi::ExitCode>() {
                                 std::process::exit(error_code.code as i32)
                             }
                         }
+                        #[cfg(not(feature = "wasi"))]
+                        RuntimeError::Error { .. } => (),
                     }
                     panic!("error: {:?}", err)
                 }

